kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala

```diff
@@ -0,0 +1,207 @@
+
+  override def shouldRunAsync: Boolean = true
+
+  private lazy val _operationLog = OperationLog.createOperationLog(session, getHandle)
+
+  private var builder: ProcBuilder = _
+
+  @volatile
+  private[kyuubi] var appIdAndUrl: Option[(String, String)] = None
+
+  private var resultFetched: Boolean = _
+
+  private val applicationCheckInterval =
+    session.sessionConf.get(KyuubiConf.BATCH_APPLICATION_CHECK_INTERVAL)
+  private val applicationCheckExecutor =
+    ThreadUtils.newDaemonSingleThreadScheduledExecutor(s"application-checker-${session.batchId}")
+
+  override def getOperationLog: Option[OperationLog] = Option(_operationLog)
+
+  override protected def beforeRun(): Unit = {
+    OperationLog.setCurrentOperationLog(_operationLog)
+    setHasResultSet(false)
+    setState(OperationState.PENDING)
+  }
+
+  override protected def afterRun(): Unit = {
+    OperationLog.removeCurrentOperationLog()
+  }
+
+  override protected def runInternal(): Unit = {
+    val asyncOperation: Runnable = () => {
+      setState(OperationState.RUNNING)
+      try {
+        submitBatchJob()
+        setState(OperationState.FINISHED)
+      } catch onError()
+    }
+    try {
+      val opHandle = session.sessionManager.submitBackgroundOperation(asyncOperation)
+      setBackgroundHandle(opHandle)
+    } catch onError("submitting batch job submission operation in background, request rejected")
+  }
+
+  private def submitBatchJob(): Unit = {
+    builder = Option(batchRequest.batchType).map(_.toUpperCase(Locale.ROOT)) match {
+      case Some("SPARK") =>
+        new SparkBatchProcessBuilder(
+          session.user,
+          session.sessionConf,
+          session.batchId,
+          batchRequest,
+          getOperationLog)
+
+      case _ =>
+        throw new UnsupportedOperationException(s"Batch type ${batchRequest.batchType} unsupported")
+    }
+
+    try {
+      info(s"Submitting ${batchRequest.batchType} batch job: $builder")
+      val process = builder.start
+      startApplicationChecker()
+      process.waitFor()
+      if (process.exitValue() != 0) {
+        throw new KyuubiException(s"Process exit with value ${process.exitValue()}")
+      }
+    } finally {
+      builder.close()
+    }
+  }
+
+  private def startApplicationChecker(): Unit = {
+    val checker = new Runnable {
+      override def run(): Unit = {
+        if (appIdAndUrl.isEmpty) {
+          try {
+            builder match {
+              case sparkBatchProcessBuilder: SparkBatchProcessBuilder =>
+                sparkBatchProcessBuilder.getApplicationIdAndUrl() match {
+                  case Some(appInfo) => appIdAndUrl = Some(appInfo)
+                  case _ =>
+                }
+
+              case _ =>
+            }
+          } catch {
+            case e: Exception => error(s"Failed to check batch application", e)
+          }
+        }
+      }
+    }
+
+    applicationCheckExecutor.scheduleAtFixedRate(
```

Comment:
why do we need to schedule it at fix rate?

Reply:
using a single thread and Thread.sleep(interval) instead?

Reply:
I guess why shall block it in the current thread in L83

Reply:
got it